### PR TITLE
Feature/reprocessor rate limiting

### PR DESCRIPTION
**Pull request by rkm**
## Proposed Changes

Adds the `set-sleep-time-ms` control message to DicomReprocessor, which can be used to control the sleep time between batches of documents.

[CHANGELOG](https://github.com/SMI/SmiServices/compare/develop...feature/reprocessor-rate-limiting#diff-4ac32a78649ca5bdd8e0ba38b7006a1e)

## Types of changes

What types of changes does your code introduce? Tick all that apply.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation-Only Update (if none of the other choices apply)
  - In this case, ensure that the message of the head commit from the source branch is prefixed with `[skip ci]`

## Checklist

By opening this PR, I confirm that I have:

- [x] Reviewed the [contributing](https://github.com/SMI/SmiServices/blob/master/CONTRIBUTING.md) guidelines for this repository
- [x] Ensured that the PR branch is in sync with the target branch (i.e. it is automatically merge-able)
- [x] Updated any relevant API documentation
- [ ] Created or updated any tests if relevant
  - Currently no tests for any of the control queue messages
- [x] Accurately updated the [CHANGELOG](https://github.com/SMI/SmiServices/blob/master/CHANGELOG.md)
  - NOTE: This ***must*** include any changes to any of the following files: default.yaml, any of the RabbitMQ server configurations, GlobalOptions.cs
- [x] Listed myself in the [CONTRIBUTORS](https://github.com/SMI/SmiServices/blob/master/CONTRIBUTORS.md) file 🚀
- [x] Requested a review by one of the repository maintainers

## Issues

- None

### CODE REVIEW

**Comment by rkm** [comment]
Note that this build did indeed pass, but was accidentally restarted and then cancelled: https://travis-ci.org/SMI/SmiServices/builds/629208695

File: src/common/Smi.Common/Options/GlobalOptions.cs [hunk 1]
```diff
@@ -273,10 +273,9 @@ public class DicomReprocessorOptions
 
         public ProducerOptions ReprocessingProducerOptions { get; set; }
 
-        public override string ToString()
-        {
-            return GlobalOptions.GenerateToString(this);
-        }
+        public TimeSpan SleepTime { get; set; }
+
+        public override string ToString() => GlobalOptions.GenerateToString(this);
```
Comment (tznind) [inline src/common/Smi.Common/Options/GlobalOptions.cs]: In future lets do this kind of thing (using statements, whitespace and => properties etc) in a seperate PR.  Makes it easier to view only the changes for the feature

File: src/microservices/Microservices.DicomReprocessor/DicomReprocessorControlMessageHandler.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+﻿
+using System;
+using JetBrains.Annotations;
+using NLog;
+using Smi.Common.Options;
+
+
+namespace Microservices.DicomReprocessor
+{
+    public class DicomReprocessorControlMessageHandler
```
Comment (tznind) [inline src/microservices/Microservices.DicomReprocessor/DicomReprocessorControlMessageHandler.cs]: Just a thought :

Can't we just make this `IControlHandler` with a method signature for `void ControlMessageHandler(string, string)` that way we can avoid Action<string,string> and make it easier to find them all etc?

File: src/common/Smi.Common/Options/GlobalOptions.cs (hunk 1 above)
Comment (rkm) [reply src/common/Smi.Common/Options/GlobalOptions.cs]: Sure, that makes sense. I think the issue is that I use the auto-formatter in VS which results in lots of diffs like that. I did investigate using [pre-commit](https://pre-commit.com/)/StyleCop before but didn't want to roll that out without chatting to everyone first.

File: src/microservices/Microservices.DicomReprocessor/DicomReprocessorControlMessageHandler.cs (hunk 2 above)
Comment (rkm) [reply src/microservices/Microservices.DicomReprocessor/DicomReprocessorControlMessageHandler.cs]: That definitely makes sense, which makes me wonder why I implemented it this way in the first place! Do we want to refactor that in this PR?

File: src/common/Smi.Common/Options/GlobalOptions.cs (hunk 1 above)
Comment (tznind) [reply src/common/Smi.Common/Options/GlobalOptions.cs]: Yeah I get a bit of that too from visual studio but I just unstage the bits that are pointless (like whitespace / using etc) and don't bother to commit them.  Takes a bit more time but helps pick up accidental code changes etc too.

File: src/microservices/Microservices.DicomReprocessor/DicomReprocessorControlMessageHandler.cs (hunk 2 above)
Comment (tznind) [reply src/microservices/Microservices.DicomReprocessor/DicomReprocessorControlMessageHandler.cs]: Yeah it should be very small fix

Comment (rkm) [reply src/microservices/Microservices.DicomReprocessor/DicomReprocessorControlMessageHandler.cs]: It's maybe something to do with the fact that not all services need specific control handlers, so they have to register a callback if they wish to handle any unique control messages.

**Pull request merged**